src/FormControl/FormControl.tsx

```diff
@@ -0,0 +1,232 @@
+      )
+    }
+    if (inputProps?.required) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        `instead of passing the 'required' prop directly to the input component, it should be passed to the parent component, <FormControl>`
+      )
+    }
+  }
+
+  if (!labelChild) {
+    // eslint-disable-next-line no-console
+    console.error(
+      `The input field with the id ${id} MUST have a FormControl.Label child.\n\nIf you want to hide the label, pass the 'visuallyHidden' prop to the FormControl.Label component.`
+    )
+  }
+
+  if (variant === 'choice') {
+    if (validationChild) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'Validation messages are not rendered for an individual checkbox or radio. The validation message should be shown for all options.'
+      )
+    }
+
+    if (React.Children.toArray(children).find(child => React.isValidElement(child) && child.props?.required)) {
+      // eslint-disable-next-line no-console
+      console.warn('An individual checkbox or radio cannot be a required field.')
+    }
+  } else {
+    // TODO: reconsider if we even need this check. The UI will just look like something is wrong
+    if (React.isValidElement(InputComponent) && (InputComponent.type === Checkbox || InputComponent.type === Radio)) {
+      // eslint-disable-next-line no-console
+      console.warn('The Checkbox or Radio components are only intended to be used with the "choice" variant')
+    }
+
+    if (
+      React.Children.toArray(children).find(
+        child => React.isValidElement(child) && child.type === FormControlLeadingVisual
+      )
+    ) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'A leading visual is only rendered for a checkbox or radio form control. If you want to render a leading visual inside of your input, check if your input supports a leading visual.'
+      )
+    }
+  }
+
+  return (
+    <Slots
+      context={{
+        captionId,
+        disabled,
+        id,
+        required,
+        validationMessageId
+      }}
+    >
+      {slots => {
+        const isLabelHidden = React.isValidElement(slots.Label) && slots.Label.props.visuallyHidden
+
+        return variant === 'choice' ? (
+          <Box display="flex" alignItems={slots.LeadingVisual ? 'center' : undefined}>
+            <Box sx={{'> input': {marginLeft: 0, marginRight: 0}}}>
+              {React.isValidElement(InputComponent) &&
+                React.cloneElement(InputComponent, {
+                  id,
+                  disabled,
+                  ['aria-describedby']: captionId
+                })}
+              {React.Children.toArray(children).filter(
+                child =>
+                  React.isValidElement(child) &&
+                  ![Checkbox, Radio].some(inputComponent => child.type === inputComponent)
+              )}
+            </Box>
+            {slots.LeadingVisual && (
```

Comment:
Not a blocker for this PR:

@mperrotti and @siddharthkp how do y'all feel about not capitalizing slot names (`slots.leadingVisual` instead of `slots.LeadingVisual`)? The PascalCase makes me think I can do something like this `<slots.LeadingVisual>` (which wouldn't work)

Reply:
that's a good point 🤔 

I call it `slots.LeadingVisual` in slots because I wanted to call it the exact same name from the component API, which is `ActionList.LeadingVisual`. 

```js
const {Slots, Slot} = createSlots(['LeadingVisual', 'InlineDescription', 'BlockDescription', 'TrailingVisual'])
```
https://github.com/primer/react/blob/main/src/ActionList2/Item.tsx#L79

Reply:
I'm cool with that change, but I'd rather make it across _all_ components instead of just this one.

Reply:
I'd like to keep `slots.LeadingVisual` until we know what changes in https://github.com/primer/react/issues/1690 in the hopes we can actually have `<Slots.LeadingVisual props="too">` 😅